lib/jsdom/living/dom-token-list.js

```diff
@@ -0,0 +1,133 @@
+"use strict";
+
+var DOMException = require("../web-idl/DOMException");
+
+// https://dom.spec.whatwg.org/#domtokenlist
+
+var INTERNAL = Symbol("DOMTokenList internal");
+
+function DOMTokenList() {
+  throw new TypeError("Illegal constructor");
+}
+
+DOMTokenList.prototype = Array.prototype;
+
+DOMTokenList.prototype.item = function item(index) {
+  return this.length <= index ? null : this[index];
+};
+
+DOMTokenList.prototype.contains = function contains(token) {
+  validateToken(token);
+  return this.indexOf(token) !== -1;
+};
+
+DOMTokenList.prototype.add = function add(/* tokens... */) {
+  var tokens = arguments;
+  var len = tokens.length;
+
+  for (var i = 0; i < len; i++) {
+    var token = tokens[i];
+    validateToken(token);
+
+    if (this.indexOf(token) === -1) {
+      this.push(token);
+    }
+  }
+  update(this);
+};
+
+DOMTokenList.prototype.remove = function remove(/* tokens... */) {
+  var tokens = arguments;
+  var len = tokens.length;
+
+  for (var i = 0; i < len; i++) {
+    var token = tokens[i];
+    validateToken(token);
+
+    var index = this.indexOf(token);
+    if (index !== -1) {
+      this.splice(index, 1);
+    }
+  }
+  update(this);
+};
+
+// if force is true, this behaves like add
+// if force is false, this behaves like remove
+// if force is undefined, this behaves as one would expect toggle to
+// always returns whether classList contains token after toggling
+DOMTokenList.prototype.toggle = function toggle(token, force) {
+  validateToken(token);
+
+  var index = this.indexOf(token);
+  if (index !== -1) {
+    if (force === false || force === undefined) {
+      this.splice(index, 1);
+      update(this);
+      return false;
+    }
+
+    return true;
+  }
+
+  if (force === false) {
+    return false;
+  }
+
+  this.push(token);
+  update(this);
+  return true;
+};
+
+DOMTokenList.prototype.toString = function toString() {
+  return this.join(" ");
+};
+
+function validateToken(token) {
+  if (token === "") {
+    throw new DOMException(DOMException.SYNTAX_ERR);
+  }
+
+  if (/\s/.test(token)) {
+    throw new DOMException(DOMException.INVALID_CHARACTER_ERR);
+  }
+}
+
+function update(list) {
+  var attribute = list[INTERNAL].attribute;
+  if (attribute !== undefined) {
```

Comment:
I have no idea! But it's part of the specification. A DOMTokenList without an attribute is a DOMSettableTokenList. Which is a thing. And not related to classList, so I didn't look into it.

I can remove it if you'd like.